aten/src/ATen/native/TensorTransformations.cpp

```diff
@@ -9,51 +10,73 @@
 namespace at {
 namespace native {
 
+template <typename scalar_t>
+void inline flip_cpu_kernel(
+  const int64_t total_dims,
+  const int64_t* stride_contiguous_d,
+  const Tensor& flip_dims_t,
+  const Tensor& in_tensor,
+  Tensor& out_tensor
+){
+  int64_t i;
+  const int64_t numel = in_tensor.numel();
+  const scalar_t* in_tensor_d = in_tensor.data<scalar_t>();
+  scalar_t* out_tensor_d = out_tensor.data<scalar_t>();
+  const int64_t* flip_dims_d = flip_dims_t.data<int64_t>();
+
+  #pragma omp parallel for private(i) if (numel > 1000)
+  for (i = 0; i < numel; i++) {
+    int64_t cur_indices = i;
+    int64_t rem = 0;
+    int64_t dst_offset = 0;
+
+    for (int64_t d = 0; d < total_dims; d++) {
+      int64_t temp = cur_indices;
+      cur_indices = cur_indices / stride_contiguous_d[d];
+      rem = temp - cur_indices * stride_contiguous_d[d];
+      for (int64_t j = 0; j < flip_dims_t.numel(); j++) {
+        if (d == flip_dims_d[j]) cur_indices = in_tensor.size(d) - 1 - cur_indices;
+      }
+      dst_offset += cur_indices * in_tensor.stride(d);
+      cur_indices = rem;
+    }
+    out_tensor_d[i] = in_tensor_d[dst_offset];
+  }
+}
+
 Tensor flip_cpu(const Tensor& self, IntList dims) {
-  const int64_t total_dims = self.dim(), flip_dims_size = dims.size();
-  flip_check_errors(total_dims, flip_dims_size, dims);
+  auto in_tensor = self;
+  const int64_t total_dims = in_tensor.dim();
 
+  dim_list_to_bitset(dims, total_dims); // returned bitset is not used, here only check correctness of dims
   auto flip_dims_v = dims.vec();
-  wrap_all_dims(flip_dims_v, total_dims);
-  std::sort(flip_dims_v.begin(), flip_dims_v.end());
-  auto final_indices = std::vector<at::Tensor>(total_dims);
-
-  auto indices = std::vector<at::Tensor>(flip_dims_size);
-  for (int64_t i = 0; i < flip_dims_size; i++) {
-    indices[i] = at::arange(self.size(flip_dims_v[i]) - 1, -1, -1, self.type().toScalarType(at::kLong));
-    // creates a meshgrid
-    auto temp = std::vector<int64_t>(flip_dims_size, 1);
-    temp[i] = indices[i].size(0);
-    indices[i] = indices[i].view(IntList(temp));
-    final_indices[flip_dims_v[i]] = indices[i];
-  }
-
-  // check if distance between two flip dims >= 2, where permute of output tensor is needed,
-  // because the advanced indexing puts all non-consecutive indices in the beginning of the tensor
-  bool to_permute = false;
-  int64_t first = flip_dims_v[0], second = flip_dims_v[0];
-  for (int64_t i = 1; i < flip_dims_size; i++) {
-    second = flip_dims_v[i];
-    if (second - first >= 2) {
-      to_permute = true;
-      break;
+  maybe_wrap_dims(flip_dims_v, total_dims);
+
+  auto sizes = in_tensor.sizes();
+  auto flip_dims_t = at::CPU(kLong).tensorFromBlob(flip_dims_v.data(), {static_cast<int64_t>(flip_dims_v.size())});
```

Comment:
why is this still a tensor?

Reply:
yeah, I just removed it. Now I am using bitset and removed inner loop at kernel. It is faster now but still slower than before:
```
>>> a = torch.randn(1000, 1000)
>>> %timeit -r 100 a.flip(0, 1)
9.68 ms ± 1.15 ms per loop (mean ± std. dev. of 100 runs, 100 loops each)
```